### DISCUSSION TITLE
Password reset doesn't work

### DISCUSSION THREAD

**Discussion by neuwoehner** [requarks/wiki#7452] [Error / Bug Report]
Wiki.js Version: 2.5.304
NodeJS Version: 16.20.2
PostgreSQL Version: 16.4

For reproduction:
Use the forgotten password function on the login page.

![image](https://github.com/user-attachments/assets/8db36623-9019-446d-ae9f-f5aabcba05cf)

Get the e-mail.

![image](https://github.com/user-attachments/assets/5e1c4484-9560-4f75-af1f-9f1d5b1a44a8)

Using password reset link in e-mail results in

![image](https://github.com/user-attachments/assets/733038c8-264a-44b4-8396-7124b67acab8)

**Comment by neuwoehner** [comment]
I was able to narrow down the error, resetting the password does not work with Chrome-based browsers, but it does with Firefox, for example.

**Comment by NGPixel** [comment]
You probably have an extension that either modifies the link target or pre-load the page rendering the token invalid.

**Comment by mritchie-kahua** [comment]
This is happening because the reset password link is a one time use link, and email scanners checking for malicious links invalidate the link before the user has a chance to click on it. Not developing anymore, but had AI run an analysis and proposed fix in hopes that could help get this moved forward. See below:

# Wiki.js Password Reset — Two-Step Reset Page Fix

## Problem

Email security scanners (Outlook Safe Links, Proofpoint, Mimecast) pre-fetch URLs in emails via GET requests. The `GET /login-reset/:token` route in Wiki.js immediately **deletes the one-time token** from the database, invalidating it before the user clicks.

## Root Cause

| File | Function | Issue |
|------|----------|-------|
| `server/controllers/auth.js` | `GET /login-reset/:token` | Calls `validateToken()` on GET — consumes token |
| `server/models/userKeys.js` | `validateToken()` | Deletes token row via `deleteById()` on first access |

## Proposed Change — Two-Step Flow

**Before:** `GET /login-reset/:token` → token consumed → render password form

**After:**

1. `GET /login-reset/:token` → token existence check only (no delete) → render **intermediate confirmation page**
2. User clicks "Reset Password" button → `POST /login-reset/:token` → token consumed → render password form

## Files to Modify

| # | File | Change |
|---|------|--------|
| 1 | `server/controllers/auth.js` | Split `GET /login-reset/:token` into GET (verify-only) + POST (consume) |
| 2 | `server/views/login-reset.pug` | **New** — intermediate page with POST form and "Reset Password" button |
| 3 | `server/controllers/auth.js` | Apply same pattern to `GET /verify/:token` (same vulnerability) |
| 4 | `server/views/verify.pug` | **New** — intermediate page for email verification (optional, same fix) |

## Detailed Changes

### `server/controllers/auth.js` — GET handler (modified)

- Query `userKeys` table for matching `resetPwd` token
- Check expiry (`validUntil`) — reject if expired
- Do **not** call `validateToken()` (which deletes)
- Render `login-reset` view passing `resetToken`

### `server/controllers/auth.js` — POST handler (new)

- Call `validateToken({ kind: 'resetPwd', token })` — this deletes the token
- Generate `changePwd` continuation token
- Render existing `login` view with `changePwdContinuationToken`

### `server/views/login-reset.pug` (new)

- Minimal branded page matching existing login styling
- Displays "Click below to reset your password"
- Single `<form method="POST">` with a submit button
- No JavaScript required — works with scanners that don't execute JS

## Security Considerations

| Concern | Mitigation |
|---------|------------|
| GET idempotency | GET now has no side effects (RFC 7231 compliant) |
| Token replay on GET | Token not consumed on GET; still single-use on POST |
| Brute force | Existing `express-brute` rate limiting preserved on both routes |
| Token expiry | 24-hour TTL unchanged; expired tokens rejected on GET |
| Scanner behavior | Scanners only issue GET, never POST with form data |
| CSRF on POST | Add CSRF token to the intermediate form |

## Testing Plan

1. Request password reset → verify email received with correct link
2. GET the link → confirm intermediate page renders, token still in DB
3. GET the link again → confirm it still works (not consumed)
4. POST via the button → confirm password change form appears, token deleted from DB
5. GET the link again → confirm it now fails (token consumed)
6. Verify `express-brute` still triggers after repeated failures
7. Verify expired tokens (>24h) are rejected on GET
